src/gameplay/glicko2.ts

```diff
@@ -0,0 +1,136 @@
+// @ts-nocheck
+
+const computeG = (phi) => {
+  return 1 / Math.sqrt(1 + (3 * phi ** 2) / Math.PI ** 2);
+};
+
+const computeE = (mu, mu_j, phi_j) => {
+  return 1 / (1 + Math.exp(-1 * computeG(phi_j) * (mu - mu_j)));
+};
+
+
+const computeNewVolatility = (delta, phi, v, sigma, tau) => {
+  const epsilon = 0.000001;
+  const a = Math.log(sigma ** 2);
+  const computeF = (x) => {
+    let output = 0;
+
+    output += Math.E ** x * (delta ** 2 - phi ** 2 - v - Math.E ** x) / (2 * (phi ** 2 + v + Math.E ** x) ** 2);
+    output -= (x - a) / tau ** 2;
+    return output;
+  };
+  
+  
+  let A = a;
+  let B = 0;
+  if (delta ** 2 > phi ** 2 + v) {
+    B = Math.log(delta ** 2 - phi ** 2 - v);
+  } else {
+    let k = 1;
+    while (computeF(a - k * tau) < 0) {
+      k += 1;
+    }
+    B = a - k * tau;
+  }
+
+
+  let f_A = computeF(A);
+  let f_B = computeF(B);
+
+  
+  while (Math.abs(B - A) > epsilon) {
+    const C = A + (A - B) * f_A / (f_B - f_A);
+    const f_C = computeF(C);
+    if (f_C * f_B <= 0) {
+      A = B;
+      f_A = f_B;
+    } else {
+      f_A /= 2;
+    }
+
+    B = C;
+    f_B = f_C;
+  }
+
+  return Math.E ** (A / 2);
+};
+
+export const glicko2Update = (playerData, matchHistory) => {
+  // Update for players who played during the rating period
+  // system constant tau is set as 0.5
+  const tau = 0.5;
+
+  // Step 2: Convert to Glicko-2 scale
+  const player = {
+    alliance: playerData.alliance,
+    mu: (playerData.request.user.playerRating - 1500) / 173.7178,
+    phi: playerData.request.user.ratingDeviation / 173.7178,
+    ratingVolatility: playerData.request.user.ratingVolatility,
+  };
+
+  // Check if the player competed during the rating period
+  if (matchHistory.length == 0) {
+    const newPhi = Math.sqrt(player.phi ** 2 + player.ratingVolatility ** 2);
+    const newRD = 173.7178 * newPhi;
+
+    return { 
+      playerRating: playerData.request.user.playerRating,
+      ratingDeviation: newRD,
+    };
+  } 
+
+  const opponents = matchHistory.map((m) => {
+    const mu = (m.opponentTeamRating - 1500) / 173.7178;
+    const phi = m.opponentTeamRatingDeviation / 173.7178;
+    return {
+      mu,
+      phi,
+      g: computeG(phi),
+      E: computeE(player.mu, mu, phi),
+      s: (m.winningTeam === player.alliance) ? 1 : 0,
+    };
+  });
+
+  // Step 3: Compute the quantity v (estimated variance)
+  let v = 0;
+  for (const opponent of opponents) {
+    const { g, E } = opponent;
```

Comment:
I prefer things fully typed out instead of abbreviations unless they're really obvious. Maybe it'll be clearer if the type of `opponent` was defined as well.